src/laddie/laddie_velocity.f90

```diff
@@ -0,0 +1,420 @@
+    TYPE(type_ice_model),                   INTENT(IN)    :: ice
+    TYPE(type_laddie_model),                INTENT(INOUT) :: laddie
+    TYPE(type_laddie_timestep),             INTENT(IN)    :: npxref
+
+    ! Local variables:
+    CHARACTER(LEN=256), PARAMETER                         :: routine_name = 'compute_viscUV'
+    INTEGER                                               :: ci, ti, tj, ei
+    REAL(dp)                                              :: D_x, D_y, D, Ah, dUabs
+    REAL(dp), DIMENSION(mesh%nTri)                        :: U_tot, V_tot
+    LOGICAL, DIMENSION(mesh%nTri)                         :: mask_oc_b_tot
+    
+    ! Add routine to path
+    CALL init_routine( routine_name)        
+
+    ! Gather
+    CALL gather_to_all_dp_1D( npxref%U, U_tot)            
+    CALL gather_to_all_dp_1D( npxref%V, V_tot)            
+    CALL gather_to_all_logical_1D( laddie%mask_oc_b, mask_oc_b_tot)
+
+    ! Loop over triangles                                  
+    DO ti = mesh%ti1, mesh%ti2
+      IF (laddie%mask_b( ti)) THEN
+    
+        ! Initialise at 0
+        laddie%viscU( ti) = 0.0_dp
+        laddie%viscV( ti) = 0.0_dp
+      
+        ! Loop over connected triangles
+        DO ci = 1, 3
+          tj = mesh%TriC( ti, ci)
+          ei = mesh%TriE( ti, ci)
+
+          D_x = mesh%Tricc( tj,1) - mesh%Tricc( ti,1)
+          D_y = mesh%Tricc( tj,2) - mesh%Tricc( ti,2)
+          D   = SQRT( D_x**2 + D_y**2)
+
+          Ah = C%laddie_viscosity ! * 0.5_dp*(SQRT(mesh%TriA( ti)) + SQRT(mesh%TriA( tj))) / 1000.0_dp
+
+          IF (tj==0) THEN
+            ! Border or corner. For now, assume no slip. If free slip: CYCLE
+            laddie%viscU( ti) = laddie%viscU( ti) - npxref%U( ti) * Ah * npxref%H_b( ti) / mesh%TriA( ti)
+            laddie%viscV( ti) = laddie%viscV( ti) - npxref%V( ti) * Ah * npxref%H_b( ti) / mesh%TriA( ti)
+          ELSE
+            ! Skip calving front - ocean connection: d/dx = d/dy = 0 
+            IF (laddie%mask_oc_b( tj)) CYCLE
+
+            dUabs = SQRT((U_tot( tj) - U_tot( ti))**2 + (V_tot( tj) - V_tot( ti))**2)
+            Ah = C%laddie_viscosity * dUabs * mesh%triCw( ti, ci) / 100.0_dp
+            
+            ! Add viscosity flux based on dU/dx and dV/dy. 
+            ! Note: for grounded neighbours, U_tot( tj) = 0, meaning this is a no slip option. Can be expanded
+            laddie%viscU( ti) = laddie%viscU( ti) + (U_tot( tj) - U_tot( ti)) * Ah * npxref%H_c( ei) / mesh%TriA( ti) * mesh%TriCw( ti, ci) / D
+            laddie%viscV( ti) = laddie%viscV( ti) + (V_tot( tj) - V_tot( ti)) * Ah * npxref%H_c( ei) / mesh%TriA( ti) * mesh%TriCw( ti, ci) / D
+          END IF
+        END DO
+
+      END IF !(laddie%mask_b( ti)
+    END DO !ti = mesh%ti1, mesh%ti2
+      
+    ! Finalise routine path
+    CALL finalise_routine( routine_name)
+    
+  END SUBROUTINE compute_viscUV
+
+  SUBROUTINE compute_divQUV_upstream( mesh, laddie, npxref, Hstar_b)
+    ! Upstream scheme
+
+    IMPLICIT NONE
```

Comment:
Not needed, module is already implicit none.

Reply:
Will remove